### DISCUSSION TITLE
MICROPYPATH expansion issue using MSYS2

### DISCUSSION THREAD

**Discussion by stinos** [micropython/micropython#9431] [Using MicroPython] [answered]
I'm tracking down an issue and it turns out it's some MSYS2/bash messing up; could someone confirm this also happens on their machine?
```
$ MICROPYPATH="~/.micropython" ./build-standard/micropython  -c "import sys;print(sys.path)"
['', '~C:/tools/msys2/msys64/.micropython']
```
I first thought something was wrong with how unix/main.c does expansion of `HOME`, but it turns out it's not the case as the expansion code isn't even being executed: the standard C `getenv("PATH")` call just returns the string as can be seen above! Cannot reproduce it when running the same executable under the debugger via gdb, cannot reproduce it with the unix port under Ubuntu, can reproduce it with the msvc port's micropython.exe but only when called from MSYS2's bash, so that's why I'm thinking it's bash? Am I doing something wrong here?

Minimal thing to reproduce:

```
#include <stdio.h>
#include <stdlib.h>

int main () {
   printf("MYVAR %s\n", getenv("MYVAR"));
   return 0;
}
```

```
$ gcc showpath.c && MYVAR="~/" ./a.exe
MYVAR ~C:/tools/msys2/msys64/
```

**Comment by karfas** [comment]
I don't think this can/needs to be fixed with micropython.
The shell/bash is responsible for expansion of ~/.

The simple workaround is to set your environment without fancy expansion characters. Even in windows there is a HOME environment, isn't it ?

While I have no idea what bash on msys2 is expected to do,  bash on my linux machine doesn't expand ~ within quotation marks at all:
```
[wet@eris ]$ echo ~
/home/wet
[wet@eris ]$ echo "~"
~
```

**Reply by stinos** [reply]
I'm aware this isn't a MicroPython issue, but I'm trying to figure out if someone knows what is going on then, i.e. whether this is some bug or some feature I can't find the documentation for

> The shell/bash is responsible for expansion of ~/

But not if it's in quotes (as you show with the echo)?

**Reply by stinos** [reply]
> Even in windows there is a HOME environment, isn't it ?

Yes and no; within MSYS2 shells that is defined, but in plain cmd.exe for instance not. Equivalent is USERPROFILE.

**Comment by stinos** [comment, marked as answer]
Asked on MSYS2 chat, it's specific to it: https://www.msys2.org/docs/filesystem-paths/#environment-variables
So basically it sees that `/` and converts it to the root path no matter what. Unexpected, but now at least I know what the issue is.
